### PR TITLE
Add checkpoint resume support to training

### LINKED ISSUE #16: Feature Request: Add Checkpoint Resume Support for Interrupted Training
Problem

Training large language models from scratch can take many hours or even days. If training is interrupted due to a power failure, system restart, GPU issues, or cloud session timeout, users may need to restart training from the beginning.

Proposed Solution

Add automatic checkpoint saving and resume functionality.

Suggested features:

- Save model checkpoints at configurable intervals
- Save optimizer state and scheduler state
- Resume training from the latest checkpoint using a command-line flag
- Display the last completed training step when resuming
- Option to keep only the most recent N checkpoints to save disk space

Benefits

- Prevents loss of training progress
- Improves usability for long training runs
- Makes the project more suitable for cloud environments
- Reduces computational cost after interruptions

### PR DESCRIPTION
**Pull request by RajvardhanPatil07**
## Summary
- add periodic checkpoint saving to the legacy train_transformer.py path
- support resuming from the latest checkpoint or a specific checkpoint path
- restore optimizer state, loss history, last completed step, and learning-rate metadata
- document the checkpoint/resume workflow and add a regression test

Closes #16

## Validation
- python3 scripts/train_transformer.py --help
- PYTHONPATH=. python3 tests/test_checkpoint_resume.py
- PYTHONPATH=. python3 tests/test_rl_math.py
- PYTHONPATH=. python3 tests/test_post_training_smoke.py
- python3 -m py_compile scripts/train_transformer.py src/models/transformer.py tests/test_checkpoint_resume.py
- git diff --check

### CODE REVIEW

File: scripts/train_transformer.py [hunk 1]
```diff
@@ -57,181 +67,454 @@ def get_peak_memory_report(device: str) -> str:
...
+        return []
+    paths = [
+        os.path.join(checkpoint_dir, name)
+        for name in os.listdir(checkpoint_dir)
+        if CHECKPOINT_RE.search(name)
+    ]
+    return sorted(paths, key=checkpoint_step)
+
+
+def resolve_resume_path(resume: str | None, checkpoint_dir: str) -> str | None:
```
[the remark this answers is not in the document]
Comment (RajvardhanPatil07) [reply scripts/train_transformer.py]: Updated the new training script annotations to use Python 3.9-compatible typing forms such as Optional, List, Tuple, and Dict instead of PEP 604 unions.

File: scripts/train_transformer.py [hunk 2]
```diff
@@ -57,181 +67,454 @@ def get_peak_memory_report(device: str) -> str:
...
+        'pytorch_version': torch.__version__,
+        'cuda_version': torch.version.cuda,
+        'lr_state': {
+            'current_lr': current_lr(optimizer),
+            'initial_lr': train_config['t_lr'],
+            'decayed_lr': train_config['t_lr_decayed'],
+            'decay_step': train_config['t_lr_decay_step'],
+        },
+    }
+    torch.save(payload, path)
```
[the remark this answers is not in the document]
Comment (RajvardhanPatil07) [reply scripts/train_transformer.py]: Updated this to write checkpoints atomically: the payload is saved to a temporary file in the checkpoint directory, then moved into place with os.replace only after torch.save succeeds. The regression test also checks that temp files are cleaned up.

**Comment by FareedKhan-dev** [comment]
Hi @RajvardhanPatil07 — thank you for this! Checkpoint save/resume was a real gap, and your implementation was clean and correct: atomic temp-file writes, `--resume [latest|path]`, LR-aware optimizer restore, prune-old-checkpoints, and a proper round-trip unit test.

We ended up with three independent PRs for the same feature (#19, #20, #28). To avoid landing conflicting checkpoint designs in `train_transformer.py`, we've consolidated on **#28**, which was iterated directly on top of your approach — it addresses the review comments raised on this PR, keeps the same atomic-write + resume semantics, and adds the Python-3.9 typing fix (`from __future__ import annotations`).

Before deciding, I validated all three branches on a 2×L40 GPU (train → checkpoint → `--resume latest` → continues from the saved step) and your branch passed. The merged design is essentially yours — thank you for the contribution and the test. 🙏

Closing in favor of #28 (now merged).
